### DISCUSSION TITLE
EloquentCollection changing to array when try to emit it as data

### DISCUSSION THREAD

**Discussion by beataburczyk** [livewire/livewire#5408] [7. Bugs (archived)] [answered]
Hi, I did select category as a bug as I believe this is a bug, but will appreciate any explanation why it's not a bug.
Also I have to mention I have found a workaround on my problem, so will post it here for someone with similar problem.

My setup:
Laravel: 9.42.2, PHP: 8.1.12, Livewire: 2.10.7

> Properties can ONLY be either JavaScript-friendly data types (string, int, array, boolean), OR one of the following PHP types: Stringable, Collection, DateTime, Model, EloquentCollection.

According to docs mentioned above EloquentCollection should be supported with no problems. Please see below to recreate a problem.

src/app/Http/Controllers/RecipeController.php
```
$recipes = Recipe::with(['preparesteps', 'tags'])->where('is_active', 1)->get();
return view('recipes.index', ['recipes' => $recipes, 'tags' => $tags]);
```
src/resources/views/recipes/index.blade.php
```
<livewire:recipe.recipes-menu :tags="$tags" />
<livewire:recipe.recipes-list :recipes="$recipes" />
```
src/resources/views/livewire/recipe/recipes-list.blade.php
```
            @foreach ($recipes as $recipe)
            {{ dump($recipe) }}
            @endforeach
```
At this point my template is returning collection as expected

> {"id":1,"name":{"en":"Accusantium laboriosam similique ut.","pl":"PL A ... cut off rest of the output...

When wire:click on item from tags, one of the tag is selected and its name is used to create new filtered collection, pretty basic stuff.
src/app/Http/Livewire/Recipe/RecipesMenu.php
```
class RecipesMenu extends Component
{
    public $tags;

    public function selectTag($name)
    {
        $recipes =  Recipe::with(['preparesteps', 'tags'])
            ->whereHas('tags', function (Builder $query) use ($name) {
                $query->where('name', '=', $name);
            })->get();

        $this->emit('refreshRecipes', $recipes);
    }

    public function render()
    {
        return view('livewire.recipe.recipes-menu');
    }
}
```

src/app/Http/Livewire/Recipe/RecipesList.php
```
class RecipesList extends Component
{
    public EloquentCollection $recipes;
    
    protected $listeners = ['refreshRecipes'];

    public function refreshRecipes($recipes)
    {
        $this->recipes = $recipes;
    }

    public function render()
    {
        return view('livewire.recipe.recipes-list', ['recipes' => $this->recipes]);
    }
}
```

When wire:click on item from tags, error indicate now conversion from collection to array happen.

> Cannot assign array to property App\Http\Livewire\Recipe\RecipesList::$recipes of type Illuminate\Database\Eloquent\Collection

Yes, I can remove typing from properties, but this will pass array anyway and result with error during a view rendering.
If Livewire supports collection why when data are passed to other component, are converted to array? I was expecting here to have persisting data type, when passing from one to another component ,via balde variables, properties or emitted data.

My workaround ;)
I have changed what I'm passing as emitted data, and which component is collecting filtered data from database.

src/app/Http/Livewire/Recipe/RecipesMenu.php
```
class RecipesMenu extends Component
{
    public $tags;

    public function selectTag($name)
    {
        $this->emit('refreshRecipes', $name);
    }

    public function render()
    {
        return view('livewire.recipe.recipes-menu');
    }
}
```

src/app/Http/Livewire/Recipe/RecipesList.php
```
class RecipesList extends Component
{
    public $recipes;
    protected $listeners = ['refreshRecipes'];

    public function refreshRecipes($name)
    {
        $this->recipes = Recipe::with(['preparesteps', 'tags'])
        ->whereHas('tags', function (Builder $query) use ($name) {
            $query->where('name', '=', $name);
        })->get();
    }

    public function render()
    {
        return view('livewire.recipe.recipes-list', ['recipes' => $this->recipes]);
    }
}
```

Sorry for long post but wanted to give some context.
Thanks in advance
Beata

**Comment by joshhanley** [comment, marked as answer]
@beataburczyk the issue is that any data that gets passed to emit, gets sent through the front end, to communicate between components. So what that means is the collection needs to get converted into javascript, which is where it is becoming an array, for the data to be sent.

I would strongly recommend not emitting your whole collection, as it has both performance and security concerns, when sending whole model data via the front end.

I would recommend only emitting an array of ID's and then listening to them on the other end and restoring them from the database that way.

Hope this helps!
